sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala

```diff
@@ -430,80 +431,127 @@ object GpuDeviceManager extends Logging {
-        val ret = confPinnedSize + (1024 * 1024 * 1024)
-        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set using " +
-            s"${RapidsConf.PINNED_POOL_SIZE} + 1 GiB instead for memory limit " +
-            s"${ret / 1024 / 1024.0} MiB")
-        ret
-      } else {
-        // We don't have pinned or a conf limit set, so go with the min
-        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set and neither is " +
-            s"${RapidsConf.PINNED_POOL_SIZE}. Using the minimum memory size for the limit " +
-            s"which is ${RapidsConf.TASK_OVERHEAD_SIZE} * cores (${GpuDeviceManager.numCores}) " +
-            s"+ 1 GiB => ${minMemoryLimit / 1024 / 1024.0} MiB")
-        minMemoryLimit
-      }
-
-      val finalMemoryLimit = if (memoryLimit < minMemoryLimit) {
-        logWarning(s"The memory limit ${memoryLimit / 1024 / 1024.0} MiB " +
-            s"is smaller than the minimum limit size ${minMemoryLimit / 1024 / 1024.0} MiB " +
-            s"using the minimum instead")
-        minMemoryLimit
       } else {
-        memoryLimit
+        // in case we cannot query the host for available memory due to environmental
+        // constraints, we can fall back to minMemoryLimit via saying there's no available
+        lazy val availableHostMemory = memCheck.getAvailableMemoryBytes(conf).getOrElse(0L)
+        val hostMemUsageFraction = .8
+        lazy val basedOnHostMemory = (hostMemUsageFraction * ((1.0 * availableHostMemory /
+          deviceCount) - heapSize - pysparkOverhead - sparkOffHeapSize)).toLong
+        if (executorOverhead.isDefined) {
+          val basedOnConfiguredOverhead = executorOverhead.get - sparkOffHeapSize
+          logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set; we derived " +
+            s"a memory limit from ($executorOverheadKey - " + s"$sparkOffHeapSizeKey) = " +
+            s"(${executorOverhead.get} - " + s"$sparkOffHeapSize) = $basedOnConfiguredOverhead")
+          if (basedOnConfiguredOverhead < minMemoryLimit) {
+            logWarning(s"memory limit $basedOnConfiguredOverhead is less than the minimum of " +
+              s"$minMemoryLimit; using the latter")
+            if (minMemoryLimit > basedOnHostMemory) {
+              logWarning(s"the amount of available memory detected on the host is " +
+                s"$availableHostMemory, based off of which we computed a limit of " +
+                s"$basedOnHostMemory, which is less than the minimum $minMemoryLimit, " +
+                s"so we are using the minimum $minMemoryLimit")
+            }
+            minMemoryLimit
+          } else {
+            basedOnConfiguredOverhead
+          }
+        } else {
+          logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set; we used " +
+            s"memory limit derived from $hostMemUsageFraction * (estimated available " +
+            s"host memory - $heapSizeKey - $pysparkOverheadKey - $sparkOffHeapSizeKey) = " +
```

Comment:
add to the message " / deviceCount" and we need to add it to the evaluated part.